mm2src/kdf_walletconnect/src/client/stream.rs

```diff
@@ -0,0 +1,293 @@
+
+                                // Perform compaction if required.
+                                if self.requests.len() * 3 < self.requests.capacity() {
+                                    self.requests.shrink_to_fit();
+                                }
+
+                                None
+                            } else {
+                                Some(StreamEvent::InboundError(ClientError::InvalidResponseId))
+                            }
+                        },
+                    }
+                },
+
+                Message::Close(frame) => {
+                    self.close_frame = frame.clone();
+                    Some(StreamEvent::ConnectionClosed(frame.clone()))
+                },
+            },
+
+            Err(error) => Some(StreamEvent::InboundError(error.into())),
+        }
+    }
+
+    fn poll_write(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), WebsocketClientError>> {
+        let mut should_flush = false;
+
+        loop {
+            // `poll_ready() needs to be called before each `start_send()` to make sure the
+            // sink is ready to accept more data.
+            match self.socket.poll_ready_unpin(cx) {
+                // The sink is ready to accept more data.
+                Poll::Ready(Ok(())) => {
+                    if let Poll::Ready(Some(next_message)) = self.outbound_rx.poll_recv(cx) {
+                        if let Err(err) = self.socket.start_send_unpin(next_message) {
+                            return Poll::Ready(Err(WebsocketClientError::TransportError(err.to_string())));
+                        }
+
+                        should_flush = true;
+                    } else if should_flush {
+                        // We've sent out some messages, now we need to flush.
+                        return self
+                            .socket
+                            .poll_flush_unpin(cx)
+                            .map_err(|err| WebsocketClientError::TransportError(err.to_string()));
+                    } else {
+                        return Poll::Pending;
+                    }
+                },
+
+                Poll::Ready(Err(err)) => {
+                    return Poll::Ready(Err(WebsocketClientError::TransportError(err.to_string())))
+                },
+
+                // The sink is not ready.
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+    }
+}
+
+impl Stream for ClientStream {
+    type Item = StreamEvent;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.is_terminated() {
+            return Poll::Ready(None);
+        }
+
+        while let Poll::Ready(data) = self.socket.poll_next_unpin(cx) {
+            match data {
+                Some(result) => {
+                    if let Some(event) =
+                        self.parse_inbound(result.map_err(|err| WebsocketClientError::TransportError(err.to_string())))
```

Comment:
nit: would be nicer if we handle the erroring case of the result here and only `parse_inbound` the valid msgs.

Reply:
done